### PR TITLE
Add a service annotation that allows endpoints for unready pods

### PR DESCRIPTION
**Pull request by bprashanth**
See https://github.com/kubernetes/kubernetes/issues/25283 for context 
@smarterclayton @thockin yes?

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/25284)

### CODE REVIEW

**Comment by smarterclayton** [comment]
I'm ok with this, although I think the right solution is to create a DNS path that always returns endpoints and a separate one that returns unready endpoints.  If most things that are clusters need this then a big chunk of services would have this annotation.

File: pkg/controller/endpoint/endpoints_controller.go [hunk 1]
```diff
@@ -53,6 +54,14 @@ const (
...
 	// avoid a hot loop, we'll wait this long between checks.
 	PodStoreSyncedPollPeriod = 100 * time.Millisecond
+
+	// An annotation on the Service denoting if the endpoints controller should
+	// go ahead and create endpoints for unready pods. This annotation is
+	// currently only used by PetSets, where we need the pet to be DNS
+	// resolvable during initialization. In this situation we create a headless
+	// service just for the PetSet, and clients shouldn't be using this Service
+	// for anything so unready endpoints don't matter.
+	tolerateUnreadyEndpointsAnnotation = "pod.alpha.kubernetes.io/tolerate-unready-endpoints"
```
Comment (smarterclayton) [inline pkg/controller/endpoint/endpoints_controller.go]: Should this be `service.alpha`?

Comment (bprashanth) [inline pkg/controller/endpoint/endpoints_controller.go]: changed

**Comment by bprashanth** [comment]
Assigning to Clayton since @thockin is at coreos fest

**Comment by smarterclayton** [comment]
Core changes look fine, add an e2e test so we can validate it.

**Comment by bprashanth** [comment]
This pr seems to be wedged from y/day, unsure why tests haven't run. I'm just going to rebase and hope they do.

**Comment by bprashanth** [comment]
@smarterclayton added e2e, tests passed, PTAL

**Comment by smarterclayton** [comment]
LGTM, I set no release note since this is really early.

**Comment by eparis** [comment]
this is a feature, why does it have P2?

**Comment by bprashanth** [comment]
Because it's blocking further work on a "p0" feature? for some definition of p0 that I didn't come up with, or apply to petset, but nevertheless has been applied

**Comment by eparis** [comment]
thanks for the explanation!

**Pull request merged**